### ISSUE TITLE
Clarification of memoization in exercise `coordinate-transformation`

### ISSUE THREAD

**Issue by SichangHe** [exercism/javascript#1782]
## Context

In the last task of exercise `coordinate-transformation`, the instruction introduces memoization as following:

>Memoizing is sometimes called dynamic programming, it allows for expensive operations to be done only once since their result is remembered.

However, the instruction also says:

> […] returns a new function that remembers the inputs to the supplied function so that the last return value can be "remembered" and only calculated once if it is called again with the same arguments.

which says that the function should only remember its last input and result.

## Problem

This is from a computer science point of view.

Dynamic programming usually memoize all previous computation. Memoizing all the previous results enables dynamic programming algorithms to compute a certain problem only once.

In this task of exercise `coordinate-transformation`, it only asks to "remember" *the last* input and result, which does not completely match the idea of dynamic programming in general.

The immediate disadvantage the above mismatch introduce could be:

- The learner confuses the concept of dynamic programming with caching or something else and meets with obstacles later when studying dynamic programming.
- The learner who knows about dynamic programming tries to do dynamic programming and fails the test so they get confused.

## Potential solutions

1. Rephrase the quote and indicate that the problem is different from general dynamic programming problem, something like:
>Memoizing is sometimes called dynamic programming, it allows for expensive operations to be done only once since their result is remembered. Note that while in dynamic programming often all previous operations are remembered, in this exercise we only want to the last one to be remembered.
2. Modify the task so it asks the learner to record all previous computations in stead of the just the last one.
3. Add yet another task which asks the learner to record all previous computations.

**Comment by junedev** [comment]
@SichangHe Thanks a lot for raising this issue with such a great description. I really appreciate it.

I agree with your assessment here. I also always had the feeling something was off with with that task but did not take the time yet to bring it out as clearly as you did.

I would recommend to implement solution 1 suggested above as a first step here to address the immediate issue quickly.

Re solution 2 and 3: Can you advise on what a good solution would look like if this would be real Memoizing? I am unsure whether handling the two coordinate parts adds to much overhead here that distracts from the actual objective of teaching closures. :thinking:

**Comment by SichangHe** [comment]
For a solution for "real" memoizing, there are multiple ways to do it but one of them is presented in the previous issue [Improve Concept Exercise: Coordinate Transformation (`closures`)](https://github.com/exercism/javascript/issues/1373#issue-1006429104) from @SleeplessByte 
under the "Solution that should not pass (but passes)" heading:

>[…] but the many many ways to build a cache make it harder to write automated analyzers for, and we really want to force people to use `let` (or `var`) here.
>## Solution that should not pass (but pass)
>```javascript
>export function memoizeTransform(f) {
>  const cache = {}
>
>  return function memoize(x, y) {
>    const key = `${x}:${y}`
>    
>    return cache[key] || cache[key] = f(x, y)
>  }
>}
>```

There are indeed many ways to "build a cache" here.
At a glance, you can have a global variable as the cache, or a variable in`memoizeTransform` as shown in this example, or a variable in `memoize`, or a variable under either function's `this`.

And, it indeed does not require using `let` or `var`, although I'm not so sure what that point is about.

**Comment by junedev** [comment]
Thanks for linking that other issue. I also don't have all the background here why it was done this way so following the [Chesterton's Fence](https://exercism.org/docs/community/being-a-good-community-member/chestertons-fence) guideline, lets stick with solution 1 then. Would you like to create a PR for that?

**Comment by SichangHe** [comment]
> Thanks for linking that other issue. I also don't have all the background here why it was done this way so following the [Chesterton's Fence](https://exercism.org/docs/community/being-a-good-community-member/chestertons-fence) guideline, lets stick with solution 1 then. Would you like to create a PR for that?

I would like to adopt solution 1 as well. For the PR, I am worried about the exact wording, so I guess we could discuss a bit first.

---

@SleeplessByte Thank you for your feedback on this.

I see the revised wording is better.

As for dropping references of dynamic programming, I am in favor of not doing so. I think the exercise introduces memoization, and it is indeed closely related to dynamic programming. Also, these kind of text adds a little sugar in the exercise.

I recognize that dynamic programming does not "usually" memoize, instead, it generally refers to bottom-up approaches. That was a mistake I made. However, in my experience, when dynamic programming does memoize, usually all previous results are remembered. But, that's debatable.

For the intended use of `let` or `var`, I don't think it would impose any real restriction. What one can do with `let`, theoretically, one could use a `const` with an object to achieve the same effect. But, I guess that is fine.

---

That said, I revised the rewording from @SleeplessByte a bit. Hopefully, it would be an enhancement.

>Memoizing is sometimes used in dynamic programming. It allows for expensive operations to be done only once since their results are remembered. **Note** that in this exercise only the last result is remembered, unlike some solutions in dynamic programming that memoize *all* results.

Grammar changed a little bit.
Using `result` instead of `computation` to hopefully reduce confusion.
Avoid indicating how much portion of dynamic programming memoizes, just say `some`.

---

If we are happy about the wording, I would like to create a PR. Otherwise, we can continue to change it.
